include/GafferScene/ExistenceQuery.h

```diff
@@ -0,0 +1,83 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2021, Cinesite VFX Ltd. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFERSCENE_EXISTENCEQUERY_H
+#define GAFFERSCENE_EXISTENCEQUERY_H
+
+#include "GafferScene/Export.h"
+#include "GafferScene/ScenePlug.h"
+#include "GafferScene/TypeIds.h"
+
+#include "Gaffer/ComputeNode.h"
+#include "Gaffer/StringPlug.h"
+#include "Gaffer/TypedPlug.h"
+
+#include <string>
+
+namespace GafferScene
+{
+
+struct GAFFERSCENE_API ExistenceQuery : Gaffer::ComputeNode
+{
+	ExistenceQuery( std::string const& name = defaultName< ExistenceQuery >() );
+	~ExistenceQuery() override;
+
+	GAFFER_NODE_DECLARE_TYPE( GafferScene::ExistenceQuery, ExistenceQueryTypeId, Gaffer::ComputeNode );
+
+	ScenePlug* scenePlug();
+	ScenePlug const* scenePlug() const;
```

Comment:
Personally I find it easier to stick to the convention that const follows the identifier it affects, ie.

void foo() const // the function name is to the left of the const
int const foo() const // the int is to the left of the const
int const foo( float const* arg ) const // the float is to the left of the const
int const foo( float const* const arg ) const // the pointer (*) is to the left of the const

The alternative is that sometimes const is to the right, sometimes its to the left

void foo() const // const is to the right
const int foo() const // const is to the right for foo and to the left for int
const int foo( const float* arg ) const // const is to the right for foo and to the left for int and float
const int foo( const float* const arg ) const // const is to the right for foo and pointer(*) and left for int and float

Reply:
Yeah, I'm aware of that viewpoint, and it is perfectly reasonable. Personally I find `T const*` confusing because visually the `const` could be attached to either left or right - in fact, visually it's _more_ attached to the right (because there is no space before `*`). In the `const T *` form it is clearly attached only to the right. The `const T *` form covers the vast majority of use cases, which means you only have to back up and force yourself into an unnatural right-to-left parsing for the rare use cases where the const-ness of the pointer is being specified. [This](https://stackoverflow.com/questions/3694630/c-const-reference-before-vs-after-type-specifier/3695979#3695979) implies I'm not alone, and that much of the standard literature (and the standard itself) prefers that form.

We could debate this forever, and both preferences are reasonable, but this is a large codebase and maintaining consistency really does help comprehension. When something doesn't fit an established pattern you have to back up and check it, which means you're not focussing on more important things. I know it's not easy adapting, but I would be grateful if we can please stick to the conventions that are already established.

Reply:
what about ```T const *``` where it is actually attached to the T ?

Reply:
> what about `T const *` where const is actually attached to the T ?

In a codebase where the convention is `const T *`, this _looks_ like it's doing something different. You have to back up and read it right to left to convince yourself that it isn't. Conventions let us _say_ something different only when we _mean_ something different, and it's then easier to spot the exceptions we _should_ be paying attention to.

> If you want both T and the pointer const you end up writing const T * const

But again, this form is less ambiguous visually. Even without knowing the right-to-left rule, you can see clearly that the first `const` is only adjacent to the `T` and the second is only adjacent to the `*`. They couldn't be referring to anything else. Whereas when you put a `const` in the middle, it's ambiguous until you recall and apply the right-to-left rule.

Like I said, I do think there are decent arguments both ways, and I'm lucky that the convention we have now is the one I've grown used to. But a codebase with conventions for this is better than one without, and conventions are only useful if we stick to them.

Reply:
ok so what about this one,

```const ConstObjectPtr object```or ```ObjectConstPtr const object``` ?

Reply:
`ConstObjectPtr` is a smart pointer to a const Object, just as `const Object *` is a raw pointer to a const Object. It's a good example because `ConstObjectPtr` is a _type_ - the convention is even embodied in the ABI.

`ObjectConstPtr` isn't a thing, because it's not a type we've defined.

`ConstObjectPtr const object` is a clever example because as you point out, it would match `const Object * const object`. But my preference would be the conventional `const ConstObjectPtr`. It avoids getting back into the original mess for an almost entirely hypothetical use of `const ConstObjectPtr *const` :)

Reply:
My point was that ```const const T*``` is not valid code, so ```const ConstObjectPtr``` is jarring due to reading const const. Also the const that applies to the pointer (Ptr) to now no where near it.

Reply:
Yeah, it's a clever example. But `ConstObjectPtr` is a type, not a `*`, and the convention is to put `const` to the left. If my noddy grepping is accurate, there are 1806 uses of `Const*Ptr` in the codebase, and 22 of `const Const` - not exactly a high proportion even if it is jarring. I suspect there are more deviations from the `const` convention in this single PR.

But again, my point isn't that your preference is without merit. My point is that conventions are necessary to share a large codebase effectively within a team, and we've chosen a convention already, for better or worse. Google told me that what we're arguing about is `const West` vs `East const`, but it's beginning to feel a bit more like the North-going Zax and the South-going Zax. Can we just please follow the existing convention and merge this? Then maybe at some point we can have a chat and see if we can `clang-format` ourselves into a happier place.

Reply:
Just interested to hear your opinion on these corner cases, happy to follow the convention.

Reply:
Great, thanks!